File: core/src/com/broll/gainea/server/core/cards/impl/play/C_TeleportGate.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.utils.PlayerUtils;
import com.broll.gainea.server.core.utils.UnitControl;

import java.util.List;
import java.util.stream.Collectors;

public class C_TeleportGate extends Card {
    public C_TeleportGate() {
        super(56, "Teleporter", "Teleportiert eine eurer Armeen auf ein beliebiges freies Land der gleichen Karte");
        setDrawChance(0.3f);
    }

    @Override
    public boolean isPlayable() {
        return true;
    }

    @Override
    protected void play() {
        Location from = selectHandler.selectLocation("Welche Armee soll teleportiert werden?", owner.getControlledLocations());
        List<Location> locations = from.getContainer().getExpansion().getAllAreas().stream().filter(Location::isFree).collect(Collectors.toList());
        Location target = selectHandler.selectLocation("Wohin?", locations);
        UnitControl.move(game, PlayerUtils.getUnits(owner, from), target);
    }

}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_TeleportGate.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.cards.EffectType.MOVEMENT
import com.broll.gainea.server.core.utils.UnitControl.move
import com.broll.gainea.server.core.utils.getUnits

class C_TeleportGate : Card(
    56,
    MOVEMENT,
    "Teleporter",
    "Teleportiert eine eurer Armeen auf ein beliebiges freies Land der gleichen Karte"
) {
    init {
        drawChance = 0.3f
    }

    override val isPlayable: Boolean
        get() = true

    override fun play() {
        val from = selectHandler.selectLocation(
            "Welche Armee soll teleportiert werden?",
            owner.controlledLocations.toList()
        )
        val locations = from.container.expansion.allAreas.filter { it.free }
        val target = selectHandler.selectLocation("Wohin?", locations)
        game.move(owner.getUnits(from), target)
    }
}
